spec.emu

```diff
@@ -676,134 +765,210 @@
+          </tr>
+          <tr>
+            <td>[[OriginalLine]]</td>
+            <td>a non-negative integer</td>
+          </tr>
+          <tr>
+            <td>[[OriginalColumn]]</td>
+            <td>a non-negative integer</td>
+          </tr>
+          <tr>
+            <td>[[NameIndex]]</td>
+            <td>a non-negative integer</td>
+          </tr>
+        </table>
+      </emu-table>
+
+      <emu-clause id="sec-DecodeMappingsField" type="sdo">
+        <h1>
+          DecodeMappingsField (
+            _state_: a Decode Mapping State Record,
+            _mappings_: a List of Decoded Mapping Records,
+            _names_: a List of Strings,
+            _sources_: a List of Decoded Source Records,
+          )
+        </h1>
+        <dl class="header"></dl>
+        <emu-grammar>
+          SegmentList :
+            Segment `;` SegmentList
+
+          MappingList :
+            Mapping `,` MappingList
+        </emu-grammar>
+        <emu-alg>
+          1. For each child node _child_ of this Parse Node, do
+            1. If _child_ is an instance of a nonterminal, then
+              1. Perform DecodeMappingsField of _child_ with arguments _state_, _mappings_, _names_ and _sources_.
+        </emu-alg>
+        <emu-grammar>
+          Segment :
+            MappingList?
+        </emu-grammar>
+        <emu-alg>
+          1. Set _state_.[[GeneratedLine]] to _state_.[[GeneratedLine]] + 1.
+          1. Set _state_.[[GeneratedColumn]] to 0.
+          1. If |MappingList| is present, then
+            1. Perform DecodeMappingsField on |MappingList| with arguments _state_, _mappings_, _names_ and _sources_.
+        </emu-alg>
+        <emu-grammar>
+          Mapping :
+            GeneratedColumn
+        </emu-grammar>
+        <emu-alg>
+          1. Perform DecodeMappingsField of |GeneratedColumn| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Let _decodedMapping_ be a new DecodedMappingRecord { [[GeneratedLine]]: _state_.[[GeneratedLine]], [[GeneratedColumn]]: _state_.[[GeneratedColumn]], [[OriginalSource]]: *null*, [[OriginalLine]]: *null*, [[OriginalColumn]]: *null*, [[Name]]: *null* }.
+          1. Append _decodedMapping_ to _mappings_.
+        </emu-alg>
+        <emu-grammar>
+          Mapping :
+            GeneratedColumn OriginalSource OriginalLine OriginalColumn Name?
+        </emu-grammar>
+        <emu-alg>
+          1. Perform DecodeMappingsField of |GeneratedColumn| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Perform DecodeMappingsField of |OriginalSource| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Perform DecodeMappingsField of |OriginalLine| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Perform DecodeMappingsField of |OriginalColumn| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. Let _source_ be _sources_[_state_.[[SourceIndex]]].
```

Comment:
It's only in an error case, but there's a small semantic difference between the old version and this I think. If the source index is invalid, the old algorithm kept the source as `null` while this returns `undefined` in JS or out-of-bounds lookup in general. (original line/column are also kept `null` when they are negative)

Reply:
Yeah there is some bound checks missing. Same for `names`.

I'm not sure about `undefined` though. We are using the [List specification type](https://tc39.es/ecma262/#sec-list-and-record-specification-type) here which does not actually spell out what is returned for out-of-bounds accesses, so it's even worse since its undefined behavior.

I'll add some bound checks.

Reply:
I rewrote the algorithm here to do the actual validation and set the fields explicitly to *null*. PTAL.

Reply:
I also filed #184. It seems some implementations (e.g. Chrome) behave as I wrote the spec initially but others (e.g. Firefox) behave as the existing spec says.